tao_compiler/mlir/disc/transforms/shape_utils.h

```diff
@@ -0,0 +1,336 @@
+
+  struct DimValue {
+    enum State : int32_t { FOLD = 0, UNFOLD = 1, INVALID = 2 } state;
+    int64_t foldVal = INT64_MIN;
+    Value unfoldVal = nullptr;
+    DimValue() { state = INVALID; }
+    explicit DimValue(Value val) {
+      unfoldVal = val;
+      state = (val == nullptr) ? INVALID : UNFOLD;
+    }
+    explicit DimValue(int64_t val) {
+      foldVal = val;
+      state = (val >= mlir::ShapedType::kDynamicSize) ? FOLD : INVALID;
+    }
+    bool isValid() const { return state != INVALID; }
+    void dump() const {
+      if (state == FOLD) {
+        llvm::errs() << "Fold value: " << foldVal << "\n";
+      } else if (state == UNFOLD) {
+        llvm::errs() << "Unfold value: ";
+        auto& non_const = const_cast<Value&>(unfoldVal);
+        non_const.dump();
+      } else {
+        llvm::errs() << "Invalid DimValue\n";
+      }
+    }
+    inline bool operator==(const DimValue& dimVal) const {
+      if (state == dimVal.state) {
+        return (state == FOLD && foldVal == dimVal.foldVal) ||
+               (state == UNFOLD && unfoldVal == dimVal.unfoldVal);
+      }
+      return false;
+    }
+    inline bool operator!=(const DimValue& dimVal) const {
+      return !(*this == dimVal);
+    }
+    inline bool operator<(const DimValue& dimVal) const {
+      if (state != dimVal.state) {
+        return state < dimVal.state;
+      } else if (foldVal != dimVal.foldVal) {
+        return foldVal < dimVal.foldVal;
+      } else {
+        return unfoldVal.getAsOpaquePointer() <
+               dimVal.unfoldVal.getAsOpaquePointer();
+      }
+    }
+    std::size_t hash() const {
+      std::size_t h = llvm::hash_value(state);
+      h = llvm::hash_combine(h, llvm::hash_value(foldVal));
+      h = llvm::hash_combine(h, mlir::hash_value(unfoldVal));
+      return h;
+    }
+  };
+
+  struct DimValueHash {
+    std::size_t operator()(const DimValue& dimVal) const {
+      return dimVal.hash();
+    }
+  };
+
+  struct DimValueContainerHash {
+    std::size_t operator()(const std::vector<DimValue>& ds) const {
+      std::size_t hash = llvm::hash_value(ds.size());
+      for (const auto& d : ds) {
+        hash = llvm::hash_combine(hash, d.hash());
+      }
+      return hash;
+    }
+  };
+
+  LogicalResult run();
+
+  Operation* getOperation() { return op_; }
+
+  Type getRefinedType(Value value);
+
+  bool isDimEqual(Value lhs, int64_t lhsDim, Value rhs, int64_t rhsDim);
+  bool isShapeEqual(Value lhs, Value rhs);
+  bool isShapeValueEqual(Value lhs, Value rhs);
+
+  // Insert tie_shape ops to explicit tie dimension equality in the IR level.
+  // Used for shape simplification pass.
+  LogicalResult buildTieShapeOps();
+
+  // Returns true if `lhs` and `rhs` are supposed to have same number of
+  // elements.
+  bool HasSameNumElements(Value lhs, Value rhs);
+
+  // Extract continuous equal dims between `lhs` and `rhs`, with the
+  // consideration of dim-linearize (e.g., caused by reshape).
+  bool extractContDimEqualInfo(
```

Comment:
Better to use continuous instead of cont.

Reply:
"Cont" is a common abbreviation. I prefer to use it to make the function name shorter.